core/src/oauth/connection.rs

```diff
@@ -338,17 +371,62 @@ impl Connection {
         Ok(())
     }
 
-    pub async fn access_token(
+    pub async fn finalize(
+        &mut self,
+        store: Box<dyn OAuthStore + Sync + Send>,
+        code: &str,
+    ) -> Result<()> {
+        // If we finalized while waiting for the lock return early.
+        match self.status {
+            // Pending is the expected status for a new connection.
+            ConnectionStatus::Pending => (),
+            // We allow calling finalized twice with the same code (user messes up or refresh).
+            // Otherwise we error.
+            ConnectionStatus::Finalized => match self.unseal_authorization_code()? {
+                Some(c) => {
+                    // If it's finalized and the code matches, we return early.
+                    if c == code {
+                        return Ok(());
+                    }
+                    Err(anyhow::anyhow!("Connection is already finalized"))?
+                }
+                // If we have a finalized connection without `authorization_code` we 500.
+                None => unreachable!(),
+            },
+        }
+
+        let rl = LockManager::new(vec![REDIS_URI.clone()]);
+
+        let lock = rl
+            .acquire_no_guard(
+                format!("oauth:{}", self.connection_id()).as_bytes(),
+                Duration::from_secs(REDIS_LOCK_TTL_SECONDS),
+            )
+            .await?;
```

Comment:
Can we have one function to create the lock so we avoid duplicating the lock name logic?

Reply:
We can't really dry it more than that as we need to be explicit about the release

Reply:
Yeah but at least abstract the key used for the lock. Given that we use the same.

Reply:
I think this is fine really